compiler/transf.nim

```diff
@@ -377,19 +377,39 @@ proc transformYield(c: PTransf, n: PNode): PNode =
           let lhs = c.transCon.forStmt[i]
           let rhs = transform(c, v)
           result.add(asgnTo(lhs, rhs))
+    elif e.kind in nkCallKinds:
+      var tmp = newTemp(c, e.typ, e.info)
+      let v = newNodeI(nkVarSection, e.info)
+      v.addVar(tmp, e)
+
+      result.add transform(c, v)
+
+      for i in 0..<c.transCon.forStmt.len - 2:
+        let lhs = c.transCon.forStmt[i]
+        let rhs = transform(c, newTupleAccess(c.graph, tmp, i))
+        result.add(asgnTo(lhs, rhs))
     else:
-      # Unpack the tuple into the loop variables
-      # XXX: BUG: what if `n` is an expression with side-effects?
       for i in 0..<c.transCon.forStmt.len - 2:
         let lhs = c.transCon.forStmt[i]
         let rhs = transform(c, newTupleAccess(c.graph, e, i))
         result.add(asgnTo(lhs, rhs))
```

Comment:
Are we sure that this never has side effects? Wouldn't it be safer to just alwas use a temporary variable (except for tuple constructors)?

Reply:
Introducing a temporary variable for all cases has some issues. So I'm still looking into it.

Reply:
It should good enough to cover `nkCallKinds + {nkTupleConstr, nkIfExpr, nkBlockExpr, nkStmtListExpr}`